docs/adr/0010-incoming-runtime-messages.md

```diff
@@ -0,0 +1,329 @@
+
+- The `executor.max_in_messages` field in the runtime descriptor is checked. If
+  it is equal to zero the method fails with `ErrIncomingMessageQueueFull`.
+
+- The incoming queue metadata structure is fetched. If it doesn't yet exist it
+  is populated with zero values.
+
+- If the value of the `size` field in the metadata structure is equal to or
+  larger than the value of the `executor.max_in_messages` field in the runtime
+  descriptor the method fails with `ErrIncomingMessageQueueFull`.
+
+- An `IncomingMessage` structure is generated based on the caller and method
+  body and the value of the `next_sequence_number` metadata field is used to
+  generate a proper key for storing it in the queue. The structure is inserted
+  into the queue.
+
+- The `size` and `next_sequence_number` fields are incremented and the updated
+  metadata is saved.
+
+### Queries
+
+This proposal adds the following new query methods in the roothash module by
+updating the `roothash.Backend` interface as follows:
+
+```golang
+type Backend interface {
+    // ... existing methods omitted ...
+
+    // HasIncomingMessages checks whether the given runtime has any incoming
+    // messages queued.
+    HasIncomingMessages(ctx context.Context, query *RuntimeQuery) (bool, error)
+
+    // IncomingMessages returns a list of queued incoming messages for the given
+    // runtime.
+    IncomingMessages(ctx context.Context, query *RuntimeQuery) ([]*IncomingMessage, error)
+}
+```
+
+### Runtime Host Protocol
+
+This proposal updates the existing host to runtime requests in the runtime host
+protocol as follows:
+
+```golang
+type RuntimeExecuteTxBatchRequest struct {
+    // ... existing fields omitted ...
+
+    // IncomingMessages are the incoming messages from the consensus layer that
+    // should be processed by the runtime in this round.
+    IncomingMessages []*IncomingMessage `json:"in_messages,omitempty"`
+}
+```
+
+### Rust Runtime Support Library
+
+This proposal updates the `transaction::Dispatcher` trait as follows:
+
+```rust
+pub trait Dispatcher: Send + Sync {
+    // ... existing unchanged methods omitted ...
+
+    /// Execute the transactions in the given batch.
+    fn execute_batch(
+        &self,
+        ctx: Context,
+        batch: &TxnBatch,
+        in_msgs: Vec<IncomingMessage>, // Added argument.
+    ) -> Result<ExecuteBatchResult, RuntimeError>;
+}
+```
+
+### Executor Processing
+
+The executor processing pipeline is changed such that pending incoming messages
+are queried at the start of each round and are then passed to the runtime via
+the runtime host protocol.
+
+### Runtime Processing
+
+The proposal requires that messages are processed by the runtime in queue order
+(e.g. on each round `InMessagesCount` messages are poped from the queue). This
+simplifies the design but the runtimes need to carefully consider how much
+resources to allocate for executing messages (vs. regular transactions) in a
+round.
```

Comment:
partially related to Bennet's idea of charging more gas for deeply queued messages, could we run into a situation where there are a lot of 'cheap' incoming messages that a runtime node perhaps prefers not to process, burying a more valuable incoming message?

incoming messages are arbitrary, and the consensus layer won't even check that they're meaningful. all we know is that the submitter used gas to enqueue it, but that gas fee would only go to the consensus nodes.

Reply:
The problem is with "a runtime node perhaps prefers not to process". Making it possible to cherry pick messages could lead to a situation where some messages are left in the queue forever. The runtime (e.g., not the node) can, however, decide to fail the message with an error instead of processing it (for example if it would exceed the resource limits or if the implied fee is too low).

Of course it needs to at least decode the message which incurs some costs.

We could also explicitly separate sent tokens into fee and tokens and the runtime could configure a min per-message fee (of course it would need to be independent of the actual message as the consensus layer should not try to decode messages).

Note that sending incoming messages should be more expensive than directly submitting transactions in the runtime.

Reply:
ah yes, message fee > transaction fee should be enough to prevent spam